### PR TITLE
Adding filter, textbox filter, and checkbox filter patterns

### LINKED ISSUE #348: Filter Panel with Faceted Filter
Create a new faceted filter PatternFly Design pattern, documenting design and interaction for the following use case / requirements:

- Users should be able to filter a page with faceted filters
- Faceted filters should not be combined with simple filters

First version of faceted filtering should support :
- a single text field
- one or more categories including multiple filter values

### LINKED ISSUE #397: Filter Bar
This is a proposal for a new Filter pattern that replaces the [existing Filter pattern](http://www.patternfly.org/pattern-library/forms-and-controls/filter/) to document generic behaviors of the filter bar that apply universally to all supported filter types.

The pattern should address the following:
* How the filter bar integrates with the overall [Toolbar](http://www.patternfly.org/pattern-library/forms-and-controls/toolbar/) when used as a part of a content view.
* How the number of results should be reflected in the filter bar (and what happens when there are no results).
* How currently applied filters are reflected and cleared (i.e. filter chips)
* How filters can be saved and recalled.

The assumption is that the Filter Bar can embed a variety of filter types (components) and that these will be documented as separate patterns and linked from the Filter Bar pattern. Related issues include:

* Filter Panel with Faceted Filter: #348 
* Toggle Filter: #363 
* Saved Filter: #306

### LINKED ISSUE #398: Basic Filter
We want to re-write the current [Filter pattern](http://www.patternfly.org/pattern-library/forms-and-controls/filter/) as a sub-component of the Filter Bar (see #397) .  Basic Filter (working title) includes a category/attribute selector and corresponding input method (text field, touchspin, datepicker, etc.) to input filter value(s).

### PR DESCRIPTION
**Pull request by cshinn**
## Description
- Restructures the filter pattern into a main Filter pattern and several smaller patterns to manipulate filters
- Closes #348
- Closes #398
- Closes #397 

## Changes
- Adds descriptions of filter behavior to Filter pattern
- Adds links to filter trigger patterns to Filter
- Adds design and overview files for Textbox and Checkbox Filter patterns
- Updated images for all filter patterns
- Adds link to filter pattern from Toolbar pattern

## Contribution Submission Checklist

- [X] A GitHub issue was created and reviewed with the PatternFly community.
- [X] The pattern design overview.md and design.md templates were used to provide all information, relevant to this design.
- [X] Any new files have been added to the folder for that pattern.
- [X] The folder, for brand new patterns, has been titled with the name of the pattern.
- [X] All images are in an "img" folder and embedded images have the correct filepath.
- [x] The overview.md, design.md and site.md files, for this design have been commited and pushed to the patternfly-design repository, for review.

### CODE REVIEW

File: pattern-library/forms-and-controls/filter/design/design.md [hunk 1]
```diff
@@ -1,53 +1,41 @@
...
-
-  1. **Attribute Selector** (optional): Contains a list of the possible attributes by which to filter.
-
-  1. **Filter Query Field:** Use when there are more than 15 possible values. The user’s filter query is entered here. The filter is activated when the user presses the “enter” key and all objects that do not have a match to this value are hidden.
-
-  1. **Filter Query Dropdown** (optional): Use the filter query [dropdown](http://www.patternfly.org/pattern-library/widgets/#dropdowns "Widgets &#124; PatternFly") when there are less than 15 possible values to allow the user to select from a known and fixed value list (e.g a list of statuses). The filter is activated when the user selects a value. Objects that do not have this value are hidden.
-
-![Image highlighting filter query field](img/filter_bar_2.png)  
+## Filter Trigger
+The filter trigger is a component that adds active filters to a page. It is located on the left side of the page, often within the [toolbar](http://www.patternfly.org/pattern-library/forms-and-controls/toolbar/). A filter trigger may be a simple text box or something much more complicated.
```
Comment (mcarrano) [inline pattern-library/forms-and-controls/filter/design/design.md]: Are there situations where we could have multiple filter triggers? I'm not sure, but if there would be only one. we should say that.

File: pattern-library/forms-and-controls/filter/design/design.md [hunk 2]
```diff
@@ -1,53 +1,41 @@
...
+Possible filter triggers include:
+- [Textbox filter](http://www.patternfly.org/pattern-library/forms-and-controls/textbox-filter/)
+- [Checkbox filter](http://www.patternfly.org/pattern-library/forms-and-controls/checkbox-filter/)
+- Advanced filter
```
Comment (mcarrano) [inline pattern-library/forms-and-controls/filter/design/design.md]: I think this is OK as a placeholder, but not sure what we call this. Would it be better to just not list any future items right now?

File: pattern-library/forms-and-controls/textbox-filter/design/design.md [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+
+  ![](img/dropdown-textbox-filter.png)
+  1. **Filter Query Dropdown** (optional): Use the filter query dropdown when the possible values for the attribute are known ahead of time and can be arranged into a fixed list (e.g a list of statuses). The filter is activated when the user selects a value. Objects that do not have this value are hidden.
+  1. **Active Filters**:  See [filter](http://www.patternfly.org/pattern-library/forms-and-controls/filter/) for a description of active filters and their behavior.
+
+  ![](img/alternate-textbox-filter.png)
+  1. **Special Query Field** (optional): Certain types of values, such as numerals, dates, and times, have special input fields that make them easier to enter. These special input fields should be used in place of the normal textbox when they exist for the selected attribute. Examples include the [touchspin](http://www.patternfly.org/pattern-library/widgets/#bootstrap-touchspin) [datepicker](http://www.patternfly.org/pattern-library/widgets/#bootstrap-datepicker) and [timepicker](http://www.patternfly.org/pattern-library/widgets/#timepicker).
```
Comment (mcarrano) [inline pattern-library/forms-and-controls/textbox-filter/design/design.md]: Missing a "," between touchspin and datepicker.

**Comment by cshinn** [comment]
@mcarrano re: the syntax hint text, they are intended to be sentence case, but the relevant attribute treated as a proper noun as they tend to be in different places. Should I change from that?

**Comment by mcarrano** [comment]
I see your point @cshinn , but we just finished changing the current filter pattern and all the content views to read more like "Filter by name" rather than "Filter by Name".  There was a long discussion about this that I don't fully recall. @LHinson please confirm and comment before Chris changes everything to conform with this.

**Comment by cshinn** [comment]
@mcarrano I've gone ahead and fixed the capitalization in a separate commit from the other changes. I can easily reverse the commit if we decide that the previous way was better.

**Comment by mcarrano** [comment]
Changes look good @cshinn . But see that you are still missing site.md files for new patterns to be added.

**Comment by Mgranfie** [comment]
@cshinn This looks good!  Just a couple of higher level issues, the design and descriptive info is clear and concise.
1. Is this going to be included as a variation on the current filter? 
2. Would you be able to provide some brief info on when to use and not to use?...

**Comment by cshinn** [comment]
@Mgranfie 
1. Which pattern are you referring to here? I believe the idea here is that we have a single pattern that talks about filters generally and their behavior and several smaller patterns that describe different ways of adding filters to a view.

2. The overview.md files contain suggestions for when to use each pattern

**Comment by Mgranfie** [comment]
@cshinn It sounds as if this is a different use/implementation/variation on the filtering UI.... 
All set with the overview.md. LGTM!

**Comment by cshinn** [comment]
@Mgranfie This would replace the documentation for the current filter pattern to make it a bit more modular, but it shouldn't affect the implementation much or at all

**Comment by mcarrano** [comment]
I am going to move ahead with merging this PR so we can proceed with other Filter stories. If anyone sees issues, we can clean up with a future story. Thanks for all of your efforts on this @cshinn ! It puts us in a much better place to extend going forward.

**Pull request merged**